app/src/main/java/com/example/weather/MainActivity.java

```diff
@@ -156,23 +156,30 @@ private String parseTemperature(String response) throws JSONException {
         return temperature;
     }
 
-    @Nullable
-    private String doRequest(URL weatherEndpoint) throws IOException {
+    void doRequest(URL weatherEndpoint, RequestCallback callback) throws IOException {
         Log.d(TAG, "doRequest start");
-        String respond = null;
+        final String respond;
 
         InputStream stream = null;
         HttpsURLConnection connection = null;
 
         try {
             connection = (HttpsURLConnection) weatherEndpoint.openConnection();
             connection.setRequestMethod("GET");
-            connection.setReadTimeout(3000);
+            connection.setReadTimeout(R.string.request_timeout);
```

Comment:
R.string.request_timeout - it is an int ID which is generated by gradle. It is not equal 3000 for sure.

Use string resources only for storing strings. Int constants should be defined as private static final int in the class.